File: leetcode/search_a_3d_array/soln.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.size() == 0 || matrix[0].size() == 0) {
            return false;
        }
        vector<int> first_column;
        for (auto row: matrix) {
            first_column.push_back(row[0]);
        }
        auto row_it = std::upper_bound(first_column.begin(), first_column.end(), target);
        if (row_it <= first_column.begin()) {
            return false;
        }
        int row_idx = row_it - first_column.begin() - 1;
        vector<int>& row = matrix[row_idx];
        return std::binary_search(row.begin(), row.end(), target);
    }
};

File: leetcode/search_a_3d_array/soln.py
import bisect

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        row_idx = bisect.bisect_right([row[0] for row in matrix], target) - 1
        if row_idx < 0:
            return False
        if target > matrix[row_idx][-1]:
            return False
        col_idx = bisect.bisect_left(matrix[row_idx], target)
        if col_idx != len(matrix[row_idx]) and matrix[row_idx][col_idx] == target:
            return True
        return False
